runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/jobsubmission/InMemoryJobServiceTest.java

```diff
@@ -186,7 +193,59 @@ public void testJobSubmissionUsesJobInvokerAndIsSuccess() throws Exception {
     verify(invocation, times(1)).start();
   }
 
+  @Test
+  public void testInvocationCleanup() {
+    final JobApi.GetJobsRequest getJobsRequest = JobApi.GetJobsRequest.newBuilder().build();
+
+    final int maxRunningJobs = maxInvocationHistory + new Random().nextInt(50);
+    // Store state listeners to be able to complete the invocations
+    final List<Consumer<JobApi.JobStateEvent>> stateListeners = new ArrayList<>(maxRunningJobs);
+
+    for (int i = 0; i < maxRunningJobs; i++) {
+      when(invocation.getId()).thenReturn(String.valueOf(i));
+      prepareAndRunJob();
+
+      // Retrieve the state listener for this invocation
+      ArgumentCaptor<Consumer<JobApi.JobStateEvent>> stateListener =
+          ArgumentCaptor.forClass(Consumer.class);
+      verify(invocation, times(i + 1)).addStateListener(stateListener.capture());
+      stateListeners.add(stateListener.getValue());
+
+      // Retrieve current list of jobs
+      RecordingObserver<JobApi.GetJobsResponse> recorder = new RecordingObserver<>();
+      service.getJobs(getJobsRequest, recorder);
+      assertThat(recorder.isSuccessful(), is(true));
+
+      // All running invocations must be available and never be discarded
+      assertThat(recorder.getValue().getJobInfoCount(), is(i + 1));
+    }
+
+    // Complete the invocations one by one and check invocation history
+    JobApi.JobStateEvent terminalEvent =
+        JobApi.JobStateEvent.newBuilder().setState(JobApi.JobState.Enum.DONE).build();
+
+    for (int i = 0; i < maxRunningJobs; i++) {
+      // finish invocation
+      stateListeners.get(i).accept(terminalEvent);
+
+      RecordingObserver<JobApi.GetJobsResponse> recorder = new RecordingObserver<>();
+      service.getJobs(getJobsRequest, recorder);
+
+      // All running invocations must never be discarded but we keep a maximum number of completed
+      // invocations
+      int jobInfoCount = recorder.getValue().getJobInfoCount();
+      if (i / maxInvocationHistory == 0) {
```

Comment:
Why not just `i < maxInvocationHistory`?

Reply:
True. I think I was feeling funny when I wrote this.